internal/operator/statefulset/builder.go

```diff
@@ -0,0 +1,871 @@
+	etcdBackupRestoreImage string
+	initContainerImage     string
+	sts                    *appsv1.StatefulSet
+	logger                 logr.Logger
+
+	clientPort int32
+	serverPort int32
+	backupPort int32
+}
+
+func newStsBuilder(client client.Client,
+	logger logr.Logger,
+	etcd *druidv1alpha1.Etcd,
+	replicas int32,
+	useEtcdWrapper bool,
+	imageVector imagevector.ImageVector,
+	sts *appsv1.StatefulSet) (*stsBuilder, error) {
+	etcdImage, etcdBackupRestoreImage, initContainerImage, err := utils.GetEtcdImages(etcd, imageVector, useEtcdWrapper)
+	if err != nil {
+		return nil, err
+	}
+	provider, err := getBackupStoreProvider(etcd)
+	if err != nil {
+		return nil, err
+	}
+	return &stsBuilder{
+		client:                 client,
+		logger:                 logger,
+		etcd:                   etcd,
+		replicas:               replicas,
+		useEtcdWrapper:         useEtcdWrapper,
+		provider:               provider,
+		etcdImage:              etcdImage,
+		etcdBackupRestoreImage: etcdBackupRestoreImage,
+		initContainerImage:     initContainerImage,
+		sts:                    sts,
+		clientPort:             pointer.Int32Deref(etcd.Spec.Etcd.ClientPort, defaultClientPort),
+		serverPort:             pointer.Int32Deref(etcd.Spec.Etcd.ServerPort, defaultServerPort),
+		backupPort:             pointer.Int32Deref(etcd.Spec.Backup.Port, defaultBackupPort),
+	}, nil
+}
+
+// Build builds the StatefulSet for the given Etcd.
+func (b *stsBuilder) Build(ctx component.OperatorContext) error {
+	b.createStatefulSetObjectMeta()
+	return b.createStatefulSetSpec(ctx)
+}
+
+func (b *stsBuilder) createStatefulSetObjectMeta() {
+	b.sts.ObjectMeta = metav1.ObjectMeta{
+		Name:            b.etcd.Name,
+		Namespace:       b.etcd.Namespace,
+		Labels:          b.getStatefulSetLabels(),
+		OwnerReferences: []metav1.OwnerReference{b.etcd.GetAsOwnerReference()},
+	}
+}
+
+func (b *stsBuilder) getStatefulSetLabels() map[string]string {
+	stsLabels := map[string]string{
+		druidv1alpha1.LabelComponentKey: common.StatefulSetComponentName,
+		druidv1alpha1.LabelAppNameKey:   b.etcd.Name,
+	}
+	return utils.MergeMaps[string, string](b.etcd.GetDefaultLabels(), stsLabels)
+}
+
+func (b *stsBuilder) createStatefulSetSpec(ctx component.OperatorContext) error {
+	podVolumes, err := b.getPodVolumes(ctx)
+	if err != nil {
+		return err
+	}
+
+	backupRestoreContainer, err := b.getBackupRestoreContainer()
+	if err != nil {
+		return err
+	}
+
+	b.sts.Spec = appsv1.StatefulSetSpec{
+		Replicas: pointer.Int32(b.replicas),
+		Selector: &metav1.LabelSelector{
+			MatchLabels: b.etcd.GetDefaultLabels(),
+		},
+		PodManagementPolicy:  defaultPodManagementPolicy,
+		UpdateStrategy:       defaultUpdateStrategy,
+		VolumeClaimTemplates: b.getVolumeClaimTemplates(),
+		ServiceName:          b.etcd.GetPeerServiceName(),
+		Template: corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels:      utils.MergeMaps[string](b.etcd.Spec.Labels, b.getStatefulSetLabels()),
```

Comment:
https://github.com/gardener/etcd-druid/pull/777#discussion_r1594210522
This is unnecesary, however if we do choose to specify type arguments to a generic function, let's try and decide a convention.
In some places, all type arguments are specified and in some places, only the first. Here only the first type argument is mentioned.